### PR TITLE
Adjust gas price limit based on earned fees

### PR DESCRIPTION
**Pull request by e00E**
With the current hardcoded maximum gas price limit we can run into
problems when gas prices are so high that the limit is not enough - even
when the submitting the solution would pay out more than the gas cost.
This commit changes the maximum gas prices to be a function of the gas
use of the solution, the fee payout and the cost of ether in the fee
(owl).

Later we probably want to estimate the price for ether in owl instead of hardcoding it.

The solver also outputs the fee in the solution json but I found it cleaner to calculate it.

### Test Plan
CI and added logging so we can see what is chosen as the limit.

### CODE REVIEW

File: core/src/driver/stablex_driver.rs
Comment (e00E) [inline core/src/driver/stablex_driver.rs:247]: Are the formula and these constants ok?

Comment (fleupold) [reply core/src/driver/stablex_driver.rs:247]: I think we should get the price of Ether in OWL from our price source. Prices of ether go out of date quickly and we already have this data available somewhere. Maybe better for a separate PR though.

File: core/src/driver/stablex_driver.rs
Comment (fleupold) [inline core/src/driver/stablex_driver.rs:244]: nit: Let's call this `subsidy_factor` or the like (extra sounds very vague). It should probably also be passed in via a config so that we can adjust it without code changes.

File: core/src/driver/stablex_driver.rs
Comment (fleupold) [inline core/src/driver/stablex_driver.rs:138]: Maybe we should collect all these constants in a common place?

File: core/src/driver/stablex_driver.rs
Comment (fleupold) [inline core/src/driver/stablex_driver.rs:131]: It's too bad that this method can fail (we could get rid of the hardcoded cap otherwise). Is this really a practical concern? I don't really see how a solution could be found for orders we don't have. But then asserting this here seems a bit far away from where this invariant is actually guaranteed. Maybe solution could have a burnt_fees field on it that we can populate when we get it and just read here?

File: core/src/driver/stablex_driver.rs
Comment (fleupold) [inline core/src/driver/stablex_driver.rs:650]: nit: Since there are multiple ways how burnt fees could be computed (we could also price all buy volumes in OWL and take 0.1% of it) I think it would be nice if the test could document in its name what ok means.

E.g. `burnt_fees_is_half_imbalance_of_reference_token_sold_and_bought`

File: core/src/driver/stablex_driver.rs
Comment (fleupold) [inline core/src/driver/stablex_driver.rs:238]: Maybe we can add a unit test that documents how much GWEI we would be willing to pay e.g. for our 50k market order.

**Review by fleupold — APPROVED** [review]
Looks good. I think the logic for computing the fee from a solution could also live closer to where the solution is created.

File: core/src/driver/stablex_driver.rs
Comment (e00E) [reply core/src/driver/stablex_driver.rs:131]: Some part of the code will have to do unwrap for this which I am ok with. I'll see if I find a more natural place for the calculation, that also has access to the original orders.

File: core/src/driver/stablex_driver.rs
Comment (e00E) [reply core/src/driver/stablex_driver.rs:247]: I agree and I also want to do this as part of a separate PR.

File: core/src/driver/stablex_driver.rs
Comment (nlordell) [inline core/src/driver/stablex_driver.rs:23]: Just a suggestion, we already get price estimates, and we can re-use the WETH price estimate for this.

File: core/src/driver/stablex_driver.rs
Comment (nlordell) [inline core/src/driver/stablex_driver.rs:21]: Maybe (in a future PR) we can estimate the gas of a transaction and use that to compute the gas cap.

**Review by nlordell — APPROVED** [review]
Looks great! Just some suggestions for future imrpovements to keep in mind.

File: core/src/driver/stablex_driver.rs
Comment (nlordell) [reply core/src/driver/stablex_driver.rs:141]: I was under the impression that the solver result output contains burnt fees. Maybe we can reuse that (if it actually does)?

Comment (e00E) [reply core/src/driver/stablex_driver.rs:141]: Yes it does but I am unsure whether we should be relying on that information since the solver output format is not documented. If we use it then we say the solver always has to output it. But it's not much work to calculate on our side.

Comment (nlordell) [reply core/src/driver/stablex_driver.rs:141]: Another way I thought of you make this method not fail is to just add together **all** sell amounts and subtract **all** buy amounts. I think you don't need to look at the order buy and sell tokens for this to work.

Additionally the BigInt to U256 conversion can be asserted or be saturating.

Comment (e00E) [reply core/src/driver/stablex_driver.rs]: Good idea. Implemented this.

File: driver/src/main.rs
Comment (e00E) [inline driver/src/main.rs:192]: Still the question what the default value should be.

Comment (nlordell) [reply driver/src/main.rs:192]: I think we can leave it as is. We can always re-configure the deployment if needed.

File: core/src/bigint_u256.rs
Comment (nlordell) [inline core/src/bigint_u256.rs:13]: :metal:

**Review by nlordell — APPROVED** [review]

**Comment by e00E** [comment]
@fleupold let me know if you still approve after the code review changes then we can merge

**Review by josojo — APPROVED** [review]
Cool thanks. I think it makes sense the way you were choosing the constants(subsidisation factors etc.)

**Review by fleupold — APPROVED** [review]
Looks great! Really good idea with computing the fee based on all token balances.

**Comment by e00E** [comment]
> Looks great! Really good idea with computing the fee based on all token balances.

That was Nick's idea btw.

**Pull request merged**
